core/src/core/mod.rs

```diff
@@ -198,12 +197,56 @@ impl BitVec {
 	}
 }
 
+/// Common errors
+#[derive(Fail, Debug)]
+pub enum Error {
+	/// Human readable represenation of amount is invalid
+	#[fail(display = "Amount string was invalid")]
+	InvalidAmountString,
+}
+
 /// Common method for parsing an amount from human-readable, and converting
 /// to internally-compatible u64
 
-pub fn amount_from_hr_string(amount: &str) -> Result<u64, ParseFloatError> {
-	let amount = amount.parse::<f64>()?;
-	Ok((amount * GRIN_BASE as f64) as u64)
+pub fn amount_from_hr_string(amount: &str) -> Result<u64, Error> {
+	// no i18n yet, make sure we use '.' as the separator
+	if amount.find(',').is_some() {
+		return Err(Error::InvalidAmountString);
+	}
+	let (grins, ngrins) = match amount.find('.') {
+		None => (
+			amount
+				.parse::<u64>()
+				.map_err(|_| Error::InvalidAmountString)?,
+			0,
+		),
+		Some(pos) => {
+			let (gs, tail) = amount.split_at(pos);
+			(parse_grins(gs)?, parse_ngrins(&tail[1..])?)
+		}
+	};
+	Ok(grins * GRIN_BASE + ngrins)
+}
+
+fn parse_grins(amount: &str) -> Result<u64, Error> {
+	if amount == "" {
+		Ok(0)
+	} else {
+		amount
+			.parse::<u64>()
+			.map_err(|_| Error::InvalidAmountString)
+	}
+}
+
+fn parse_ngrins(amount: &str) -> Result<u64, Error> {
```

Comment:
Why not just right-pad with zeroes and then parse as u64?

Reply:
Mostly to not allocate a tmp string. I can try to implement both (actually 3 approaches) and write benchmarks